students/alex_skrn/Lesson07Assignment/RELATIONALMAILROOM/mailroom.py

```diff
@@ -0,0 +1,438 @@
+#!/usr/bin/env python3
+
+"""Mailroom - Lesson 7 Adv Python - using a relational database.
+
+To use the mailroom module, a db must be already set up. It should now be in
+directory Lesson07Assignment - mailroom.db. If not, first run models.py
+(which will create an empty db and will be enough for the mailroom to work) and
+then add_data.py to include several donors and donations as an example.
+"""
+
+import os
+import datetime
+import tkinter as tk
+from tkinter import filedialog
+import peewee
+from models import Person, Donation
+
+
+####################
+# SINGLE DONOR CLASS
+####################
+class SingleDonor():
+    """Provide a class for a single donor."""
+
+    def __init__(self, _name, _donations):
+        """Instantiate a SingleDonor class object."""
+        self._name = _name
+        if isinstance(_donations, list):
+            self._donations = _donations
+        elif isinstance(_donations, tuple):
+            self._donations = list(_donations)
+        else:
+            self._donations = [_donations]
+
+    @property
+    def name(self):
+        """Provide a getter method for the name property."""
+        return self._name
+
+    @property
+    def donations(self):
+        """Provide a getter method for the donations property."""
+        return self._donations
+
+    def sort_by_total(self):
+        """Provide a sort_key for sorting by total donations."""
+        return sum(self._donations)
+
+    def sort_by_name(self):
+        """Provide a sort_key for sorting by name."""
+        return self._name
+
+    def __str__(self):
+        """Return self._name."""
+        return self._name
+
+    def __repr__(self):
+        """Return SingleDonor("Name", [donations])."""
+        if len(self._donations) == 1:
+            return 'SingleDonor("{}", {})'.format(self._name,
+                                                  self._donations[0]
+                                                  )
+        else:
+            return 'SingleDonor("{}", {})'.format(self._name,
+                                                  self._donations
+                                                  )
+
+    def __eq__(self, other):
+        """Return True if names and donations are the same."""
+        return (self._name, self._donations) == (other.name, other.donations)
+
+    def __lt__(self, other):
+        """Provide __lt__ method used in sorting somehow, I guess."""
+        return ((self._name, self._donations) <
+                (other.name, other.donations)
+                )
+
+    def add_donation(self, amount):
+        """Add a donation."""
+        # Reminder: Model.create() means 'instantiate-then-save'
+        Donation.create(person_name=self.name, donation=amount)
```

Comment:
no save here?

Reply:
Hi Micheal, well, no save(), I guess. From peewee docs I undersatnd that there are two alternative ways to store data. First, uncle_bob = Person(name='Bob'); uncle_bob.save(). Second, Person.create(name='Grandma'). The docs also say: "The create() method is a shorthand for instantiate-then-save." I already don't remember why I had to use this second method, probably because save() didn't always work. Thanks for your review by the way!